lib/mllparty/connection_hub/client_monitor.ex

```diff
@@ -0,0 +1,160 @@
+defmodule MLLParty.ConnectionHub.ClientMonitor do
+  use GenServer
+
+  @moduledoc """
+    Module to proactively monitor a client's connection and reconnect as necessary
+    Send slack notification if connection goes down for too long, send another when it comes back up
+  """
+
+  require Logger
+
+  alias MLLParty.ConnectionHub.ClientWrapper
+
+  def child_spec(ip, port, client_wrapper_pid) do
+    %{
+      id: "#{ip}:#{port} monitor",
+      start: {__MODULE__, :start_link, [ip, port, client_wrapper_pid]}
+    }
+  end
+
+  @impl true
+  def init([ip, port, client_wrapper_pid]) do
+    Logger.info("Starting client monitor: #{ip}:#{port}")
+    schedule_monitor()
+    # Return :ok and initial state
+    {:ok,
+     %{
+       ip: ip,
+       port: port,
+       client_wrapper_pid: client_wrapper_pid,
+       down_at: nil,
+       down_notification_sent: false,
+       reconnect_attempted: false
+     }}
+  end
+
+  def start_link(ip, port, client_wrapper_pid) do
+    GenServer.start_link(__MODULE__, [ip, port, client_wrapper_pid],
+      name: String.to_atom("#{ip}:#{port} monitor")
+    )
+  end
+
+  @impl true
+  def handle_info(:monitor, state) do
+    %{
+      endpoint: endpoint,
+      connected: connected,
+      pending_reconnect: pending_reconnect
+    } = ClientWrapper.client_status(state.client_wrapper_pid)
+
+    # Check the connection status
+    new_state =
+      case connected == true do
+        true ->
+          # Connection is up
+          handle_connection_up(endpoint, state)
+
+        false ->
+          # Connection is down
+          handle_connection_down(endpoint, connected, pending_reconnect, state)
+      end
+
+    schedule_monitor()
+    new_state = Map.merge(state, new_state)
+    {:noreply, new_state}
+  end
+
+  def handle_connection_up(endpoint, state) do
+    # Get current timestamp
+    now = DateTime.utc_now()
+    # Connection is up, check if it was down previously
+    if state.down_at do
+      seconds = DateTime.diff(now, state.down_at)
+
+      message =
+        "#{endpoint} connection is up. Last down duration: #{TimeUtils.format_seconds(seconds)}."
+
+      Logger.info(message)
+      # If a downtime notification was sent, follow up that it's connected again
+      if state.down_notification_sent do
+        SlackNotification.send_notification(message)
+      end
+    end
+
+    # Reset down_at, down_notification_sent, reconnect_attempted
+    %{down_at: nil, down_notification_sent: false, reconnect_attempted: false}
+  end
+
+  def handle_connection_down(endpoint, connected, pending_reconnect, state) do
```

Comment:
This function is made quite complex by having 2 commingled responsibilities: managing/sending slack notification & attempting to reconnect.

If I were to refactor this I'd probably start with simply breaking those 2 responsibilities into separate functions (and building on the composability theme in the other comment): `maybe_send_notification(state) -> state` and `attempt_reconnection(state) -> state`. Then `handle_connection_down` becomes quite readable:

```elixir
def handle_connection_down(state) do
  state
  |> maybe_send_notification()
  |> attempt_reconnection()
end
```

I would maybe even think about moving the notification tracking/sending to a separate process– that would allow for the `state` here to be simplified quite a bit, and would also allow the slack webhook request to occur concurrently with the reconnection attempt (currently this blocks on the webhook request prior to reconnection attempt).

Reply:
That would be easier to follow!